File: src/domain/services/viewerTracker.go
package services

import (
	"log"
	"net"
	"net/http"
	"path"
	"strconv"
	"strings"
	"sync"
	"time"

	"Theatrum/constants"
	"Theatrum/domain/models"
	"Theatrum/domain/repositories"
)

// ViewerTracker tracks concurrent viewers and total views per stream.
type ViewerTracker struct {
	mu      sync.RWMutex
	streams map[string]*streamTracking // trackingKey -> tracking data
	storage repositories.StoragePort
}

type viewerSession struct {
	startedAt  time.Time
	lastActive time.Time
}

type viewSession struct {
	startedAt  time.Time
	lastActive time.Time
	counted    bool // true once counted as a view
}

type streamTracking struct {
	mu sync.RWMutex

	// Concurrent viewers (live only)
	viewersEnabled bool
	viewerWindow   time.Duration
	activeViewers  map[string]*viewerSession // clientIP -> session

	// Total views (all types)
	viewsEnabled bool
	viewWindow   time.Duration
	viewSessions map[string]*viewSession // clientIP -> session
	totalViews   int64
	dirty        bool // true when totalViews changed since last save
}

func NewViewerTracker(storage repositories.StoragePort) *ViewerTracker {
	vt := &ViewerTracker{
		streams: make(map[string]*streamTracking),
		storage: storage,
	}
	go vt.cleanupLoop()
	return vt
}

// TrackSegmentRequest is called on every .ts request. Updates viewer activity
// and increments view count on new sessions.
func (vt *ViewerTracker) TrackSegmentRequest(trackingKey, clientIP string, viewersCfg models.Viewers, viewsCfg models.Views) {
	st := vt.getOrCreateStream(trackingKey, viewersCfg, viewsCfg)

	now := time.Now()
	st.mu.Lock()
	defer st.mu.Unlock()

	// Track concurrent viewers
	if st.viewersEnabled {
		vs, exists := st.activeViewers[clientIP]
		if !exists || now.Sub(vs.lastActive) >= st.viewerWindow {
			vs = &viewerSession{startedAt: now}
			st.activeViewers[clientIP] = vs
		}
		vs.lastActive = now
	}

	// Track views
	if st.viewsEnabled {
		sess, exists := st.viewSessions[clientIP]
		if !exists || now.Sub(sess.lastActive) >= st.viewWindow {
			sess = &viewSession{startedAt: now}
			st.viewSessions[clientIP] = sess
		}
		sess.lastActive = now
		if !sess.counted && now.Sub(sess.startedAt) >= st.viewWindow {
			st.totalViews++
			st.dirty = true
			sess.counted = true
		}
	}
}

// GetViewerCount returns the number of active viewers for a stream.
func (vt *ViewerTracker) GetViewerCount(trackingKey string) int {
	vt.mu.RLock()
	st, ok := vt.streams[trackingKey]
	vt.mu.RUnlock()
	if !ok {
		return 0
	}

	now := time.Now()
	st.mu.RLock()
	defer st.mu.RUnlock()

	count := 0
	for _, vs := range st.activeViewers {
		if now.Sub(vs.startedAt) >= st.viewerWindow && now.Sub(vs.lastActive) < st.viewerWindow {
			count++
		}
	}
	return count
}

// GetViewCount returns the total accumulated views for a stream.
// Falls back to reading from disk when the stream is not in memory.
func (vt *ViewerTracker) GetViewCount(trackingKey string) int64 {
	vt.mu.RLock()
	st, ok := vt.streams[trackingKey]
	vt.mu.RUnlock()
	if !ok {
		return vt.loadViewCount(trackingKey)
	}

	st.mu.RLock()
	defer st.mu.RUnlock()
	return st.totalViews
}

// StreamStats holds viewer and view counts for a single stream.
type StreamStats struct {
	TrackingKey string
	Viewers     int
	Views       int64
}

// GetAllStreamStats returns stats for all active streams.
// Used by the Prometheus collector to export metrics on each scrape.
func (vt *ViewerTracker) GetAllStreamStats() []StreamStats {
	vt.mu.RLock()
	keys := make([]string, 0, len(vt.streams))
	for k := range vt.streams {
		keys = append(keys, k)
	}
	vt.mu.RUnlock()

	now := time.Now()
	stats := make([]StreamStats, 0, len(keys))

	for _, key := range keys {
		vt.mu.RLock()
		st, ok := vt.streams[key]
		vt.mu.RUnlock()
		if !ok {
			continue
		}

		st.mu.RLock()
		viewers := 0
		if st.viewersEnabled {
			for _, vs := range st.activeViewers {
				if now.Sub(vs.startedAt) >= st.viewerWindow && now.Sub(vs.lastActive) < st.viewerWindow {
					viewers++
				}
			}
		}
		views := st.totalViews
		st.mu.RUnlock()

		stats = append(stats, StreamStats{
			TrackingKey: key,
			Viewers:     viewers,
			Views:       views,
		})
	}

	return stats
}

// UnregisterStream cleans up tracking data when a live stream ends.
// Persists the final view count to disk before removing from memory.
func (vt *ViewerTracker) UnregisterStream(trackingKey string) {
	vt.mu.Lock()
	st, ok := vt.streams[trackingKey]
	if ok {
		delete(vt.streams, trackingKey)
	}
	vt.mu.Unlock()

	if ok && st.viewsEnabled {
		st.mu.RLock()
		count := st.totalViews
		st.mu.RUnlock()
		vt.saveViewCount(trackingKey, count)
	}
}

func (vt *ViewerTracker) getOrCreateStream(trackingKey string, viewersCfg models.Viewers, viewsCfg models.Views) *streamTracking {
	vt.mu.RLock()
	st, ok := vt.streams[trackingKey]
	vt.mu.RUnlock()
	if ok {
		return st
	}

	vt.mu.Lock()
	defer vt.mu.Unlock()

	// Double-check after acquiring write lock
	if st, ok := vt.streams[trackingKey]; ok {
		return st
	}

	st = &streamTracking{
		viewersEnabled: viewersCfg.Enabled,
		viewerWindow:   time.Duration(viewersCfg.Window) * time.Second,
		activeViewers:  make(map[string]*viewerSession),
		viewsEnabled:   viewsCfg.Enabled,
		viewWindow:     time.Duration(viewsCfg.Window) * time.Second,
		viewSessions:   make(map[string]*viewSession),
		totalViews:     vt.loadViewCount(trackingKey),
	}
	vt.streams[trackingKey] = st
	return st
}

// cleanupLoop periodically removes expired viewer entries.
func (vt *ViewerTracker) cleanupLoop() {
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for range ticker.C {
		vt.mu.RLock()
		keys := make([]string, 0, len(vt.streams))
		for k := range vt.streams {
			keys = append(keys, k)
		}
		vt.mu.RUnlock()

		type dirtyEntry struct {
			key   string
			count int64
		}
		var toSave []dirtyEntry

		now := time.Now()
		for _, key := range keys {
			vt.mu.RLock()
			st, ok := vt.streams[key]
			vt.mu.RUnlock()
			if !ok {
				continue
			}

			st.mu.Lock()
			for ip, vs := range st.activeViewers {
				if now.Sub(vs.lastActive) >= st.viewerWindow {
					delete(st.activeViewers, ip)
				}
			}
			for ip, sess := range st.viewSessions {
				if now.Sub(sess.lastActive) >= st.viewWindow {
					delete(st.viewSessions, ip)
				}
			}
			if st.dirty && st.viewsEnabled {
				toSave = append(toSave, dirtyEntry{key: key, count: st.totalViews})
				st.dirty = false
			}
			st.mu.Unlock()
		}

		// Save dirty view counts outside of locks
		for _, entry := range toSave {
			vt.saveViewCount(entry.key, entry.count)
		}
	}
}

// viewsFilePath returns the disk path for persisted view counts.
func viewsFilePath(trackingKey string) string {
	return path.Join(constants.VideoDir, trackingKey, constants.ViewsFile)
}

// loadViewCount reads a persisted view count from disk. Returns 0 if not found.
func (vt *ViewerTracker) loadViewCount(trackingKey string) int64 {
	data, err := vt.storage.ReadFile(viewsFilePath(trackingKey))
	if err != nil {
		return 0
	}
	count, err := strconv.ParseInt(strings.TrimSpace(string(data)), 10, 64)
	if err != nil {
		return 0
	}
	return count
}

// saveViewCount writes a view count to disk.
func (vt *ViewerTracker) saveViewCount(trackingKey string, count int64) {
	if err := vt.storage.WriteFile(viewsFilePath(trackingKey), []byte(strconv.FormatInt(count, 10))); err != nil {
		log.Printf("Error persisting view count for %s: %v", trackingKey, err)
	}
}

// GetClientIP extracts the client IP from an HTTP request.
// Checks X-Forwarded-For header first for reverse proxy support.
func GetClientIP(r *http.Request) string {
	// Check X-Forwarded-For header
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		// Take the first IP (original client)
		if idx := strings.IndexByte(xff, ','); idx != -1 {
			return strings.TrimSpace(xff[:idx])
		}
		return strings.TrimSpace(xff)
	}

	// Fall back to RemoteAddr (strip port)
	ip, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		return r.RemoteAddr
	}
	return ip
}

File: src/main/java/com/fjourdren/theatrum/domain/service/ViewerTracker.java
package com.fjourdren.theatrum.domain.service;

import com.fjourdren.theatrum.application.port.in.TrackViewerUseCase;
import com.fjourdren.theatrum.application.port.out.StoragePort;
import com.fjourdren.theatrum.domain.constant.VideoConstants;
import com.fjourdren.theatrum.domain.model.AppPaths;
import com.fjourdren.theatrum.domain.model.StreamStats;
import com.fjourdren.theatrum.domain.model.Viewers;
import com.fjourdren.theatrum.domain.model.Views;
import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Tracks concurrent viewers and total views per stream.
 *
 * <p>Both counters use a delayed window: a client IP is only counted once it has been requesting
 * segments continuously for {@code window} seconds. Total views are persisted to disk so they
 * survive restarts.
 */
@Component
@Slf4j
public class ViewerTracker implements TrackViewerUseCase {

    private static final long CLEANUP_INTERVAL_SECONDS = 10;

    private final Map<String, StreamTracking> streams = new ConcurrentHashMap<>();
    private final StoragePort storage;
    private final AppPaths appPaths;
    private final Clock clock;

    private ScheduledExecutorService cleanupExecutor;

    @Autowired
    public ViewerTracker(StoragePort storage, AppPaths appPaths) {
        this(storage, appPaths, Clock.systemUTC());
    }

    ViewerTracker(StoragePort storage, AppPaths appPaths, Clock clock) {
        this.storage = storage;
        this.appPaths = appPaths;
        this.clock = clock;
    }

    @PostConstruct
    public void startCleanupLoop() {
        cleanupExecutor = Executors.newSingleThreadScheduledExecutor(
                Thread.ofVirtual().name("viewer-tracker-cleanup").factory());
        cleanupExecutor.scheduleAtFixedRate(() -> {
            try {
                runCleanupOnce();
            } catch (RuntimeException e) {
                log.error("Viewer tracker cleanup failed", e);
            }
        }, CLEANUP_INTERVAL_SECONDS, CLEANUP_INTERVAL_SECONDS, TimeUnit.SECONDS);
    }

    @PreDestroy
    public void stopCleanupLoop() {
        if (cleanupExecutor != null) {
            cleanupExecutor.shutdownNow();
        }
    }

    /**
     * Called on every segment request. Refreshes viewer activity and counts a view once the
     * client has been watching for the configured window.
     */
    @Override
    public void trackSegmentRequest(String trackingKey, String clientIp, Viewers viewersCfg, Views viewsCfg) {
        StreamTracking st = getOrCreateStream(trackingKey, viewersCfg, viewsCfg);
        Instant now = clock.instant();

        synchronized (st) {
            if (st.viewersEnabled) {
                Session session = st.activeViewers.get(clientIp);
                if (session == null || elapsed(session.lastActive, now, st.viewerWindow)) {
                    session = new Session(now);
                    st.activeViewers.put(clientIp, session);
                }
                session.lastActive = now;
            }

            if (st.viewsEnabled) {
                Session session = st.viewSessions.get(clientIp);
                if (session == null || elapsed(session.lastActive, now, st.viewWindow)) {
                    session = new Session(now);
                    st.viewSessions.put(clientIp, session);
                }
                session.lastActive = now;
                if (!session.counted && elapsed(session.startedAt, now, st.viewWindow)) {
                    st.totalViews++;
                    st.dirty = true;
                    session.counted = true;
                }
            }
        }
    }

    /** Number of clients currently watching a stream. */
    @Override
    public int getViewerCount(String trackingKey) {
        StreamTracking st = streams.get(trackingKey);
        if (st == null) {
            return 0;
        }
        Instant now = clock.instant();
        synchronized (st) {
            return countViewers(st, now);
        }
    }

    /**
     * Total accumulated views for a stream. Falls back to the persisted count when the stream is
     * not tracked in memory.
     */
    @Override
    public long getViewCount(String trackingKey) {
        StreamTracking st = streams.get(trackingKey);
        if (st == null) {
            return loadViewCount(trackingKey);
        }
        synchronized (st) {
            return st.totalViews;
        }
    }

    /** Stats for all tracked streams. Used by the Prometheus collector on each scrape. */
    @Override
    public List<StreamStats> getAllStreamStats() {
        Instant now = clock.instant();
        List<StreamStats> stats = new ArrayList<>(streams.size());

        streams.forEach((key, st) -> {
            synchronized (st) {
                stats.add(new StreamStats(key, st.viewersEnabled ? countViewers(st, now) : 0, st.totalViews));
            }
        });

        return stats;
    }

    /** Drops tracking data when a stream ends, persisting the final view count first. */
    @Override
    public void unregisterStream(String trackingKey) {
        StreamTracking st = streams.remove(trackingKey);
        if (st == null || !st.viewsEnabled) {
            return;
        }

        long count;
        synchronized (st) {
            count = st.totalViews;
        }
        saveViewCount(trackingKey, count);
    }

    /** Expires stale sessions and flushes view counts changed since the last run. */
    void runCleanupOnce() {
        Instant now = clock.instant();
        Map<String, Long> toSave = new LinkedHashMap<>();

        streams.forEach((key, st) -> {
            synchronized (st) {
                st.activeViewers.values().removeIf(session -> elapsed(session.lastActive, now, st.viewerWindow));
                st.viewSessions.values().removeIf(session -> elapsed(session.lastActive, now, st.viewWindow));
                if (st.dirty && st.viewsEnabled) {
                    toSave.put(key, st.totalViews);
                    st.dirty = false;
                }
            }
        });

        // Save dirty view counts outside of the stream locks.
        toSave.forEach(this::saveViewCount);
    }

    private StreamTracking getOrCreateStream(String trackingKey, Viewers viewersCfg, Views viewsCfg) {
        return streams.computeIfAbsent(trackingKey,
                key -> new StreamTracking(viewersCfg, viewsCfg, loadViewCount(key)));
    }

    private static int countViewers(StreamTracking st, Instant now) {
        int count = 0;
        for (Session session : st.activeViewers.values()) {
            if (elapsed(session.startedAt, now, st.viewerWindow) && !elapsed(session.lastActive, now, st.viewerWindow)) {
                count++;
            }
        }
        return count;
    }

    private static boolean elapsed(Instant since, Instant now, Duration window) {
        return Duration.between(since, now).compareTo(window) >= 0;
    }

    private Path viewsFilePath(String trackingKey) {
        return appPaths.videoDir().resolve(trackingKey).resolve(VideoConstants.VIEWS_FILE);
    }

    /** Reads a persisted view count. Returns 0 when missing or unreadable. */
    private long loadViewCount(String trackingKey) {
        try {
            return Long.parseLong(new String(storage.readFile(viewsFilePath(trackingKey)), StandardCharsets.UTF_8).trim());
        } catch (Exception e) {
            return 0;
        }
    }

    private void saveViewCount(String trackingKey, long count) {
        try {
            storage.writeFile(viewsFilePath(trackingKey), Long.toString(count).getBytes(StandardCharsets.UTF_8));
        } catch (Exception e) {
            log.error("Error persisting view count for " + trackingKey + ": " + e);
        }
    }

    /** Per-stream tracking state. All fields are guarded by the instance monitor. */
    private static final class StreamTracking {
        private final boolean viewersEnabled;
        private final Duration viewerWindow;
        private final Map<String, Session> activeViewers = new HashMap<>();

        private final boolean viewsEnabled;
        private final Duration viewWindow;
        private final Map<String, Session> viewSessions = new HashMap<>();

        private long totalViews;
        private boolean dirty;

        private StreamTracking(Viewers viewersCfg, Views viewsCfg, long totalViews) {
            this.viewersEnabled = viewersCfg.enabled();
            this.viewerWindow = Duration.ofSeconds(viewersCfg.window());
            this.viewsEnabled = viewsCfg.enabled();
            this.viewWindow = Duration.ofSeconds(viewsCfg.window());
            this.totalViews = totalViews;
        }
    }

    /** A single client's watching session. {@code counted} is only used for view sessions. */
    private static final class Session {
        private final Instant startedAt;
        private Instant lastActive;
        private boolean counted;

        private Session(Instant startedAt) {
            this.startedAt = startedAt;
            this.lastActive = startedAt;
        }
    }
}
